File: lib/network/packets/server/s_level_up_result.dart
class SLevelUpResult {
  const SLevelUpResult({
    required this.success,
    required this.reason,
    required this.message,
    required this.realmLevel,
    required this.exp,
    required this.expMax,
  });

  final bool success;
  final String reason;
  final String message;
  final int realmLevel;
  final int exp;
  final int expMax;

  factory SLevelUpResult.fromData(Map<String, dynamic> data) {
    return SLevelUpResult(
      success: data['success'] as bool? ?? false,
      reason: data['reason'] as String? ?? '',
      message: data['message'] as String? ?? '',
      realmLevel: data['realmLevel'] as int? ?? 1,
      exp: data['exp'] as int? ?? 0,
      expMax: data['expMax'] as int? ?? 0,
    );
  }
}

File: src/main/java/com/xin/server/packet/server/S_LevelUpResult.java
package com.xin.server.packet.server;

import com.xin.server.model.instance.PcInstance;
import com.xin.server.packet.ServerBasePacket;
import com.xin.server.packet.ServerOpcodes;

/**
 * 境界內升級結果。
 */
public class S_LevelUpResult extends ServerBasePacket {

    public static final String REASON_OK = "OK";
    public static final String REASON_NOT_IN_GAME = "NOT_IN_GAME";
    public static final String REASON_INVALID_AMOUNT = "INVALID_AMOUNT";
    public static final String REASON_MAX_LEVEL = "MAX_LEVEL";

    private S_LevelUpResult(boolean success, String reason, String message) {
        super(ServerOpcodes.S_LEVEL_UP_RESULT);
        put("success", success);
        put("reason", reason);
        put("message", message);
    }

    public static S_LevelUpResult ok(PcInstance pc) {
        S_LevelUpResult packet = new S_LevelUpResult(true, REASON_OK, "升級成功");
        packet.put("realmLevel", pc.getRealmLevel());
        packet.put("exp", pc.getExp());
        packet.put("expMax", pc.getExpMax());
        return packet;
    }

    public static S_LevelUpResult fail(String reason, String message) {
        return new S_LevelUpResult(false, reason, message);
    }
}
